File: config/editor/src/app/definition/platform/MqttConfig.ts
import {PlatformConfig} from "./PlatformConfig";
import {PlatformType} from "../enums/PlatformType";
import {HaDiscoveryConfig} from "./HaDiscoveryConfig";
import {Expose, Transform} from "class-transformer";
import {skipDefault} from "../AppConfig";


function createDefault(value: any) {
    if (value === undefined) {
        return new HaDiscoveryConfig();
    }
    return new HaDiscoveryConfig(value.enabled!, value.topic_prefix!);
}

export class MqttConfig extends PlatformConfig {
    host: string;
    username: string;
    password: string;
    @Transform(({value}) => skipDefault(value, 1883), {toPlainOnly: true})
    port: number = 1883;
    @Expose({name: "topic_prefix"})
    @Transform(({value}) => skipDefault(value, "boneio"), {toPlainOnly: true})
    topicPrefix: string = "boneio";
    @Expose({name: "ha_discovery"})
    @Transform(({value}) => skipDefault(value, new HaDiscoveryConfig()), {toPlainOnly: true})
    @Transform(({value}) => createDefault(value), {toClassOnly: true})
    haDiscovery: HaDiscoveryConfig = new HaDiscoveryConfig();

    public constructor(host: string, username: string, password: string) {
        super(PlatformType.MQTT);
        this.host = host;
        this.username = username;
        this.password = password;
    }

}

File: config/config-sbc2ha/src/main/java/com/dfi/sbc2ha/config/sbc2ha/definition/platform/MqttConfig.java
package com.dfi.sbc2ha.config.sbc2ha.definition.platform;

import com.dfi.sbc2ha.config.sbc2ha.definition.enums.PlatformType;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;

@Data
public class MqttConfig extends PlatformConfig {
    public static final String BONE_IO = "boneIO";

    String host;

    String username;

    String password;

    int port = 1883;

    @JsonProperty("topic_prefix")
    String topicPrefix = BONE_IO;

    @JsonProperty("ha_discovery")
    HaDiscoveryConfig haDiscovery = new HaDiscoveryConfig();

    public MqttConfig() {
        platform = PlatformType.MQTT;
    }

    @Data
    public static class HaDiscoveryConfig {
        public static final String HOMEASSISTANT = "homeassistant";

        boolean enabled = true;

        @JsonProperty("topic_prefix")
        String topicPrefix = HOMEASSISTANT;
    }
}
